pkg/validation/usage_groups.go

```diff
@@ -0,0 +1,160 @@
+// This file is a modified copy of the usage groups implementation in Mimir:
+//
+// https://github.com/grafana/mimir/blob/0e8c09f237649e95dc1bf3f7547fd279c24bdcf9/pkg/ingester/activeseries/custom_trackers_config.go#L48
+
+package validation
+
+import (
+	"fmt"
+
+	amlabels "github.com/prometheus/alertmanager/pkg/labels"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/labels"
+
+	phlaremodel "github.com/grafana/pyroscope/pkg/model"
+)
+
+const (
+	// Maximum number of usage groups that can be configured (per tenant).
+	maxUsageGroups = 50
+)
+
+var (
+	// This is a duplicate of distributor_received_decompressed_bytes, but with
+	// usage_group as a label.
+	usageGroupReceivedDecompressedBytes = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "pyroscope",
+			Name:      "usage_group_received_decompressed_total",
+			Help:      "The total number of decompressed bytes per profile received by usage group.",
+		},
+		[]string{"type", "tenant", "usage_group"},
+	)
+
+	// This is a duplicate of discarded_bytes_total, but with usage_group as a
+	// label.
+	UsageGroupDiscardedBytes = promauto.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "pyroscope",
+			Name:      "usage_group_discarded_bytes_total",
+			Help:      "The total number of bytes that were discarded by usage group.",
+		},
+		[]string{"reason", "tenant", "usage_group"},
+	)
+)
+
+type usageGroupEntry struct {
+	Name     string
+	Matchers []*labels.Matcher
+}
+
+// UsageGroupConfig is an allowlist of service names that have per-app usage
+// enabled. This allowlist is constructed on a per-tenant basis.
+type UsageGroupConfig struct {
+	tenantID string
+	config   []usageGroupEntry
+}
+
+// GetUsageGroupName matches the label set to a usage group. If no usage group
+// is matched, the default group name is used.
+func (u *UsageGroupConfig) GetUsageGroup(lbls phlaremodel.Labels) UsageGroup {
+	group := UsageGroup{
+		tenantID: u.tenantID,
+		name:     "other",
+	}
+
+	for _, cfgGroup := range u.config {
+		if matchesAll(cfgGroup.Matchers, lbls) {
+			group.name = cfgGroup.Name
+		}
+	}
+	return group
+}
+
+type UsageGroup struct {
+	tenantID string
+	name     string
+}
+
+func (u UsageGroup) CountReceivedBytes(profileType string, n int64) {
+	usageGroupReceivedDecompressedBytes.WithLabelValues(profileType, u.tenantID, u.name).Add(float64(n))
+}
+
+func (u UsageGroup) CountDiscardedBytes(reason string, n int64) {
+	UsageGroupDiscardedBytes.WithLabelValues(reason, u.tenantID, u.name).Add(float64(n))
+}
+
+// DistributorUsageGroups returns the usage groups that are enabled for this
+// tenant.
+func (o *Overrides) DistributorUsageGroups(tenantID string) (*UsageGroupConfig, error) {
```

Comment:
Any returned error in this method will lead to profiling traffic being dropped. This is not ideal.
This seems quite a severe reaction to "I have wrong accounting usage groups". 

I do suggest we should move all those errors into the parsing part of the overrides.

Reply:
Separately this method does a lot and is in the hot path of ingestion. We should also move as much as we can into the "loading the overrides" part of the code.

E.g. the parsing of label matchers should only happen only when the override is parsed and not for every profile ingested.